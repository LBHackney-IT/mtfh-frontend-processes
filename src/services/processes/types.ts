export enum Trigger {
  CloseProcess = "CloseProcess",
  CancelProcess = "CancelProcess",
}
