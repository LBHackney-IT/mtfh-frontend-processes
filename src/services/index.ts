export { default as locale } from "./locale";
export { default as menu } from "./menu";
export * as processes from "./processes";
export * from "./config";
