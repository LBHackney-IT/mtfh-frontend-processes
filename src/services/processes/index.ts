export * from "./sole-to-joint";
export * from "./change-of-name";
