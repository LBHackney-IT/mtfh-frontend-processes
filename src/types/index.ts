import { Process } from "@mtfh/common/lib/api/process/v1";

export enum Recommendation {
  Approve = "Approve",
  Decline = "Decline",
  Appointment = "Appointment",
}
export type RecommendationType = "Approve" | "Decline" | "Appointment";

export type TargetType = "tenure" | "property" | "person";

export interface ProcessSideBarProps {
  process: any;
  submitted: boolean;
  closeCase: boolean;
  setCloseProcessDialogOpen: any;
  setCancel: any;
}

export interface ProcessComponentProps {
  processConfig?: IProcess;
  process: Process;
  mutate: any;
  optional?: any;
  setGlobalError?: any;
}

export interface IStartProcess {
  thirdPartyCondition?: string;
  thirdPartyComponent?: React.FC;
  subHeading?: string;
  subComponent?: React.FC;
}

export interface IProcess {
  processName: string;
  title: string;
  targetType: TargetType;
  startProcess: IStartProcess;
  states: {
    [state: string]: {
      state: string;
      status: string;
      triggers: Record<string, string>;
    };
  };
}
