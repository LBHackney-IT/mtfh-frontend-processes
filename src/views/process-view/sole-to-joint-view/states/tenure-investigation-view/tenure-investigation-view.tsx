import React from "react";

import { ContactDetails } from "../../../../../components";
import { TenureInvestigationForm } from "../../../shared/tenure-investigation-form";

import { Process } from "@mtfh/common/lib/api/process/v1";
import { Text } from "@mtfh/common/lib/components";

interface TenureInvestigationViewProps {
  process: Process;
  mutate: () => void;
  setGlobalError: any;
  optional?: any;
}

export const TenureInvestigationView = ({
  process,
  mutate,
  setGlobalError,
  optional,
}: TenureInvestigationViewProps): JSX.Element => {
  const { tenant } = optional;

  return (
    <div data-testid="soletojoint-tenure-investigation">
      {tenant ? (
        <ContactDetails fullName={tenant.fullName} personId={tenant.id} />
      ) : (
        <Text>Tenant not found.</Text>
      )}
      <TenureInvestigationForm
        process={process}
        mutate={mutate}
        setGlobalError={setGlobalError}
      />
    </div>
  );
};
