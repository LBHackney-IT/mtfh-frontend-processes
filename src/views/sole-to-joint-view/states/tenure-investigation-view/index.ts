export * from "./tenure-investigation-view";
