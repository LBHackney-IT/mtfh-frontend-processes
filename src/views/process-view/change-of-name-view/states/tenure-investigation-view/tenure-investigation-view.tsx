import { Form, Formik } from "formik";

import {
  TenureInvestigationFormData,
  tenureInvestigationSchema,
} from "../../../../../schemas/tenure-investigation";
import { locale } from "../../../../../services";
import { Trigger } from "../../../../../services/processes/types";
import { ProcessComponentProps, Recommendation } from "../../../../../types";
import { SubmitCaseView } from "../../../shared/submit-case-view";

import { editProcess } from "@mtfh/common/lib/api/process/v1";
import {
  Button,
  Center,
  Checkbox,
  Field,
  FormGroup,
  InlineField,
  Radio,
  RadioGroup,
  Spinner,
} from "@mtfh/common/lib/components";
import { useErrorCodes } from "@mtfh/common/lib/hooks";

const { views } = locale;

export const TenureInvestigationView = ({
  processConfig,
  process,
  mutate,
  optional,
  setGlobalError,
}: ProcessComponentProps): JSX.Element => {
  const { submitted } = optional;
  const errorMessages = useErrorCodes();

  if (!errorMessages) {
    return (
      <Center>
        <Spinner />
      </Center>
    );
  }

  if (
    processConfig?.states.applicationSubmitted.state === process.currentState.state &&
    submitted
  ) {
    return (
      <SubmitCaseView
        processConfig={processConfig}
        process={process}
        mutate={mutate}
        optional={optional}
      />
    );
  }

  return (
    <div data-testid="changeofname-tenure-investigation">
      <Formik<TenureInvestigationFormData>
        initialValues={{
          tenureInvestigationRecommendation: "",
          completed: false,
        }}
        validateOnBlur={false}
        validateOnChange={false}
        validationSchema={tenureInvestigationSchema(errorMessages)}
        onSubmit={async (values) => {
          try {
            await editProcess({
              id: process.id,
              processName: process?.processName,
              etag: process.etag || "",
              processTrigger: Trigger.TenureInvestigation,
              formData: {
                tenureInvestigationRecommendation:
                  values.tenureInvestigationRecommendation.toLowerCase(),
              },
              documents: [],
            });
            mutate();
          } catch (e: any) {
            setGlobalError(e.response?.status || 500);
          }
        }}
      >
        {({ dirty, errors }) => (
          <Form noValidate style={{ marginTop: 65 }}>
            <Field
              id="tenure-investigation-recommendation"
              name="tenureInvestigationRecommendation"
              label={views.tenureInvestigation.recommendedOutcome}
              type="radio"
              required
            >
              <RadioGroup>
                <Radio
                  id="tenure-investigation-recommendation-approve"
                  value={Recommendation.Approve}
                >
                  {views.tenureInvestigation.approve}
                </Radio>
                <Radio
                  id="tenure-investigation-recommendation-appointment"
                  value={Recommendation.Appointment}
                >
                  {views.tenureInvestigation.appointment}
                </Radio>
                <Radio
                  id="tenure-investigation-recommendation-decline"
                  value={Recommendation.Decline}
                >
                  {views.tenureInvestigation.decline}
                </Radio>
              </RadioGroup>
            </Field>

            <FormGroup id="tenure-investigation-completed" error={errors.completed}>
              <InlineField name="completed" type="checkbox">
                <Checkbox id="tenure-investigation-completed-checkbox">
                  {views.tenureInvestigation.tenureInvestigationCompleted}
                </Checkbox>
              </InlineField>
            </FormGroup>

            <Button type="submit" style={{ marginTop: 65, width: 222 }} disabled={!dirty}>
              {locale.confirm}
            </Button>
          </Form>
        )}
      </Formik>
    </div>
  );
};
