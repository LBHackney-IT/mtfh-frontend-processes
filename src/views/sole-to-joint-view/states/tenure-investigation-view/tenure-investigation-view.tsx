import React, { useState } from "react";
import { Link as RouterLink } from "react-router-dom";

import { isPast } from "date-fns";

import { SoleToJointHeader } from "../../../../components";
import { AppointmentDetails } from "../../../../components/appointment-details/appointment-details";
import { AppointmentForm } from "../../../../components/appointment-form/appointment-form";
import { locale } from "../../../../services";
import { Trigger } from "../../../../services/processes/types";
import { IProcess } from "../../../../types";
import { EligibilityChecksPassedBox, TenantContactDetails } from "../shared";

import { Process, editProcess } from "@mtfh/common/lib/api/process/v1";
import { useTenure } from "@mtfh/common/lib/api/tenure/v1";
import {
  Box,
  Button,
  Center,
  Checkbox,
  ErrorSummary,
  Heading,
  Link,
  Spinner,
  StatusErrorSummary,
  StatusHeading,
  Text,
} from "@mtfh/common/lib/components";

const { views } = locale;

interface TenureInvestigationViewProps {
  processConfig: IProcess;
  process: Process;
  mutate: () => void;
}

export const TenureInvestigationView = ({
  processConfig,
  process,
  mutate,
}: TenureInvestigationViewProps): JSX.Element => {
  const isCurrentState = (state) => state === process.currentState.state;
  const [globalError, setGlobalError] = useState<number>();
  const [completed, setCompleted] = useState<boolean>(false);
  const { data: tenure, error } = useTenure(process.targetId);
  const {
    applicationSubmitted,
    hoApprovalPassed,
    tenureAppointmentScheduled,
    tenureAppointmentRescheduled,
  } = processConfig.states;
  const [needAppointment, setNeedAppointment] = useState<boolean>(
    isCurrentState(hoApprovalPassed.state),
  );
  const formData = process.currentState.processData.formData as {
    appointmentDateTime: string;
  };

  if (error) {
    return (
      <ErrorSummary
        id="tenure-investigation-view"
        title={locale.errors.unableToFetchRecord}
        description={locale.errors.unableToFetchRecordDescription}
      />
    );
  }

  if (!tenure) {
    return (
      <Center>
        <Spinner />
      </Center>
    );
  }

  const stateConfig = processConfig.states.applicationSubmitted;
  const tenant = tenure?.householdMembers.find((m) => m.isResponsible);
  const submit = async (recommendation: string) => {
    try {
      await editProcess({
        id: process.id,
        processName: process?.processName,
        etag: process.etag || "",
        processTrigger: stateConfig.triggers.tenureInvestigation,
        formData: {
          tenureInvestigationRecommendation: recommendation,
        },
        documents: [],
      });
      mutate();
    } catch (e: any) {
      setGlobalError(e.response?.status || 500);
    }
  };

  return (
    <div data-testid="soletojoint-TenureInvestigation">
      <SoleToJointHeader processConfig={processConfig} process={process} />
      {globalError && (
        <StatusErrorSummary id="tenure-investigation-global-error" code={globalError} />
      )}
      <EligibilityChecksPassedBox />
      <Box variant="success">
        <StatusHeading
          variant="success"
          title={views.submitCase.supportingDocumentsApproved}
        />
        <div
          style={{ marginLeft: 60, marginTop: 17.5 }}
          className="govuk-link lbh-link lbh-link--no-visited-state"
        >
          <Link as={RouterLink} to="#" variant="link">
            {views.reviewDocuments.viewInDes}
          </Link>
        </div>
      </Box>
      {[
        hoApprovalPassed.state,
        tenureAppointmentScheduled.state,
        tenureAppointmentRescheduled.state,
      ].includes(process.currentState.state) && (
        <>
          {/*TODO*/}
          <Box variant="success">
            <StatusHeading
              variant="success"
              title={views.tenureInvestigation.tenureInvestigatorRecommendation(
                "approve",
              )}
            />
            <div
              style={{ marginLeft: 60, marginTop: 17.5 }}
              className="govuk-link lbh-link lbh-link--no-visited-state"
            >
              <Link as={RouterLink} to="#" variant="link">
                {views.tenureInvestigation.showReport}
              </Link>
            </div>
          </Box>
          {/*TODO*/}

          <Box variant="success">
            <StatusHeading
              variant="success"
              title={views.tenureInvestigation.hoApproved}
            />
          </Box>

          <Heading variant="h2">{views.tenureInvestigation.hoApprovedNextSteps}</Heading>
          {isCurrentState(hoApprovalPassed.state) && (
            <Text>{views.tenureInvestigation.mustMakeAppointment}</Text>
          )}

          {[
            tenureAppointmentScheduled.state,
            tenureAppointmentRescheduled.state,
          ].includes(process.currentState.state) && (
            <AppointmentDetails
              process={process}
              needAppointment={needAppointment}
              setNeedAppointment={setNeedAppointment}
              options={{
                requestAppointmentTrigger: Trigger.ScheduleTenureAppointment,
                rescheduleAppointmentTrigger: Trigger.RescheduleTenureAppointment,
                appointmentRequestedState: tenureAppointmentScheduled.state,
                appointmentRescheduledState: tenureAppointmentRescheduled.state,
                cancelProcess: true,
              }}
            />
          )}
        </>
      )}

      {tenant ? <TenantContactDetails tenant={tenant} /> : <Text>Tenant not found.</Text>}

      {[
        hoApprovalPassed.state,
        tenureAppointmentScheduled.state,
        tenureAppointmentRescheduled.state,
      ].includes(process.currentState.state) && (
        <AppointmentForm
          process={process}
          mutate={mutate}
          needAppointment={needAppointment}
          setGlobalError={setGlobalError}
          setNeedAppointment={setNeedAppointment}
          options={{
            buttonText: "Continue",
            requestAppointmentTrigger: Trigger.ScheduleTenureAppointment,
            rescheduleAppointmentTrigger: Trigger.RescheduleTenureAppointment,
            appointmentRequestedState: tenureAppointmentScheduled.state,
            appointmentRescheduledState: tenureAppointmentRescheduled.state,
          }}
        />
      )}

      {[tenureAppointmentScheduled.state, tenureAppointmentRescheduled.state].includes(
        process.currentState.state,
      ) &&
        !needAppointment && (
          <Button disabled={!isPast(new Date(formData.appointmentDateTime))}>
            {views.tenureInvestigation.documentsSigned}
          </Button>
        )}

      {isCurrentState(applicationSubmitted.state) && (
        <>
          <Checkbox
            id="tenure-investigation-completed"
            checked={completed}
            onChange={() => setCompleted(!completed)}
          >
            {views.tenureInvestigation.tenureInvestigationCompleted}
          </Checkbox>
          <Text>{views.tenureInvestigation.recommendation}</Text>
          <Button
            type="submit"
            disabled={!completed}
            onClick={() => submit("approve")}
            style={{ width: 222, marginRight: "100%" }}
          >
            {views.tenureInvestigation.approve}
          </Button>
          <Button
            type="submit"
            disabled={!completed}
            onClick={() => submit("appointment")}
            style={{ width: 222, marginRight: "100%" }}
          >
            {views.tenureInvestigation.appointment}
          </Button>
          <Button
            type="submit"
            disabled={!completed}
            onClick={() => submit("decline")}
            style={{ width: 222, marginRight: "100%" }}
          >
            {views.tenureInvestigation.decline}
          </Button>
        </>
      )}
    </div>
  );
};
