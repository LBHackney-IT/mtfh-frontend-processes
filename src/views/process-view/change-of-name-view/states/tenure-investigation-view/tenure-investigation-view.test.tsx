import {
  getContactDetailsV2,
  getReferenceDataV1,
  mockContactDetailsV2,
  mockProcessV1,
  patchProcessV1,
  render,
  server,
} from "@hackney/mtfh-test-utils";
import { screen, waitForElementToBeRemoved } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { locale, processes } from "../../../../../services";
import { Trigger } from "../../../../../services/processes/types";
import { Recommendation } from "../../../../../types";
import { TenureInvestigationView } from "./tenure-investigation-view";

import * as processV1 from "@mtfh/common/lib/api/process/v1/service";
import commonLocale from "@mtfh/common/lib/locale";

let submitted = false;
let closeCase = false;
const setSubmitted = jest.fn();
const setCloseCase = () => {};

const options = {
  url: "/processes/changeofname/e63e68c7-84b0-3a48-b450-896e2c3d7735",
  path: "/processes/:processName/:processId",
};

describe("tenure-investigation-view", () => {
  beforeEach(() => {
    jest.resetModules();
    submitted = false;
    closeCase = false;
  });

  test("it renders TenureInvestigation view correctly for ApplicationSubmitted state, submitted=false", async () => {
    const editProcessSpy = jest.spyOn(processV1, "editProcess");
    server.use(getReferenceDataV1({}, 200));
    server.use(getContactDetailsV2(mockContactDetailsV2));
    const { container } = render(
      <TenureInvestigationView
        processConfig={processes.changeofname}
        process={{
          ...mockProcessV1,
          currentState: { ...mockProcessV1.currentState, state: "ApplicationSubmitted" },
        }}
        mutate={() => {}}
        optional={{ submitted, setSubmitted, closeCase, setCloseCase }}
      />,
      options,
    );
    await waitForElementToBeRemoved(screen.queryAllByText(/Loading/));
    await userEvent.click(screen.getByLabelText(Recommendation.Appointment));
    await userEvent.click(
      screen.getByLabelText(
        locale.views.tenureInvestigation.tenureInvestigationCompleted,
      ),
    );
    await userEvent.click(screen.getByText(locale.confirm));
    expect(editProcessSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        processTrigger: Trigger.TenureInvestigation,
        formData: {
          tenureInvestigationRecommendation: Recommendation.Appointment.toLowerCase(),
        },
      }),
    );
    expect(container).toMatchSnapshot();
  });

  test("it renders TenureInvestigation view correctly for ApplicationSubmitted state, submitted=true", async () => {
    server.use(getReferenceDataV1({}, 200));
    server.use(getContactDetailsV2(mockContactDetailsV2));
    const { container } = render(
      <TenureInvestigationView
        processConfig={processes.changeofname}
        process={{
          ...mockProcessV1,
          currentState: { ...mockProcessV1.currentState, state: "ApplicationSubmitted" },
        }}
        mutate={() => {}}
        optional={{ submitted: true, setSubmitted, closeCase, setCloseCase }}
      />,
      options,
    );
    await waitForElementToBeRemoved(screen.queryAllByText(/Loading/));
    expect(container).toMatchSnapshot();
    await userEvent.click(screen.getByText("Continue"));
    expect(setSubmitted).toHaveBeenCalledWith(false);
  });

  test("it renders error if submit fails", async () => {
    server.use(patchProcessV1(null, 500));
    render(
      <TenureInvestigationView
        processConfig={processes.changeofname}
        process={{
          ...mockProcessV1,
          currentState: { ...mockProcessV1.currentState, state: "ApplicationSubmitted" },
        }}
        mutate={() => {}}
        optional={{ submitted, setSubmitted, closeCase, setCloseCase }}
      />,
      options,
    );
    await waitForElementToBeRemoved(screen.queryAllByText(/Loading/));
    await userEvent.click(screen.getByLabelText(Recommendation.Appointment));
    await userEvent.click(
      screen.getByLabelText(
        locale.views.tenureInvestigation.tenureInvestigationCompleted,
      ),
    );
    await userEvent.click(screen.getByText(locale.confirm));
    await expect(
      screen.findByText(commonLocale.components.statusErrorSummary.statusTitle(500), {
        exact: false,
      }),
    ).resolves.toBeInTheDocument();
  });
});
