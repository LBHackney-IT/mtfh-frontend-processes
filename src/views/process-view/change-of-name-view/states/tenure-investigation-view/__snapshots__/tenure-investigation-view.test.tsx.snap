// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tenure-investigation-view it renders TenureInvestigation view correctly for ApplicationSubmitted state, submitted=false 1`] = `
<div>
  <div
    data-testid="changeofname-tenure-investigation"
  >
    <form
      action="#"
      novalidate=""
      style="margin-top: 65px;"
    >
      <fieldset
        class="govuk-form-group lbh-form-group"
        id="tenure-investigation-recommendation"
      >
        <legend
          class="govuk-label lbh-label"
          for="tenure-investigation-recommendation-field"
        >
          What is the recommended outcome?
          <sup
            aria-hidden="true"
          >
            *
          </sup>
        </legend>
        <div
          class="govuk-radios lbh-radios"
          id="tenure-investigation-recommendation-field"
        >
          <div
            class="govuk-radios__item"
          >
            <input
              class="govuk-radios__input"
              id="tenure-investigation-recommendation-approve"
              name="tenureInvestigationRecommendation"
              required=""
              type="radio"
              value="Approve"
            />
            <label
              class="govuk-label govuk-radios__label"
              for="tenure-investigation-recommendation-approve"
            >
              Approve
            </label>
          </div>
          <div
            class="govuk-radios__item"
          >
            <input
              class="govuk-radios__input"
              id="tenure-investigation-recommendation-appointment"
              name="tenureInvestigationRecommendation"
              required=""
              type="radio"
              value="Appointment"
            />
            <label
              class="govuk-label govuk-radios__label"
              for="tenure-investigation-recommendation-appointment"
            >
              Appointment
            </label>
          </div>
          <div
            class="govuk-radios__item"
          >
            <input
              class="govuk-radios__input"
              id="tenure-investigation-recommendation-decline"
              name="tenureInvestigationRecommendation"
              required=""
              type="radio"
              value="Decline"
            />
            <label
              class="govuk-label govuk-radios__label"
              for="tenure-investigation-recommendation-decline"
            >
              Decline
            </label>
          </div>
        </div>
      </fieldset>
      <div
        class="govuk-form-group lbh-form-group"
        id="tenure-investigation-completed"
      >
        <label
          class="govuk-label lbh-label"
          for="tenure-investigation-completed-field"
        >
          
        </label>
        <div
          class="govuk-checkboxes__item"
        >
          <input
            class="govuk-checkboxes__input"
            id="tenure-investigation-completed-field"
            name="completed"
            type="checkbox"
            value="true"
          />
          <label
            class="govuk-label govuk-checkboxes__label"
            for="tenure-investigation-completed-field"
          >
            I confirm that the tenure investigation has been completed
          </label>
        </div>
      </div>
      <button
        class="govuk-button lbh-button"
        style="margin-top: 65px; width: 222px;"
        type="submit"
      >
        Confirm
      </button>
    </form>
  </div>
</div>
`;

exports[`tenure-investigation-view it renders TenureInvestigation view correctly for ApplicationSubmitted state, submitted=true 1`] = `
<div>
  <h2
    class="lbh-heading-h2"
  >
    Next Steps
    :
  </h2>
  <p
    class="lbh-body-m"
  >
    Based upon the documents submitted by the tenant and the proposed joint tenant, and the investigation undertaken by the Housing Officer, the application now needs to be assessed by the Tenancy Investigation Team.
  </p>
  <button
    class="govuk-button lbh-button"
    style="width: 180px; margin-right: 100%;"
    type="button"
  >
    Continue
  </button>
</div>
`;
